File: programmers/lev1/42889.py
def solution(N, stages):
    answer = []
    size=len(stages)
    people=[0]*(N+2)
    for i in stages:
        people[i]+=1
    temp=[]
    print(people)
    minus=0
    for i in range(1,len(people)-1):
        if size-minus<=0:
            temp.append((0,i))
            continue
        temp.append((people[i]/(size-minus),i))
        size=size-minus
        minus=people[i]
        
    print(temp)
    temp.sort(key=lambda x:(-x[0],x[1]))
    print(temp)
    for i in temp:
        answer.append(i[1])
    if len(temp)<N:
        temp+=([0]*(N-len(answer)))
    print(answer)
    return answer

N=5
stages=[2, 1, 2, 6, 2, 4, 3, 3]
N=3
stages=[2,2,2,2,2]
solution(N,stages)

File: java/src/programmers/year2023/lv1/_42889.java
package programmers.year2023.lv1;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

public class _42889 {
    public static void main(String[] args) {
//        int N = 5;
//        int[] stages ={2, 1, 2, 6, 2, 4, 3, 3};

//        int N = 4;
//        int[] stages ={4,4,4,4,4};

        int N = 2;
        int[] stages ={1};
        solution(N, stages);
    }

    public static int[] solution(int N, int[] stages) {
        int[] answer = new int[N];
        int size = stages.length;
        HashMap<Integer, Integer> map = new HashMap<>();

        for (int i = 1; i <= N; i++) {
            map.put(i, 0);
        }

        for (int stage : stages) {
            map.put(stage, map.get(stage) + 1);
        }
        ArrayList<Failure> answerList = new ArrayList<>();
        for (int i = 0; i < N; i++) {

            answerList.add(new Failure(size != 0 ? (float) map.get(i + 1)/size : 0,i + 1));
            size -= map.get(i + 1);
        }
        Collections.sort(answerList);
        for (int i = 0; i < answerList.size(); i++) {
            Failure failure = answerList.get(i);
            answer[i] = failure.stage;
            System.out.println("i: " + i + ", stage = " + failure.stage);
            System.out.println("failure.percent = " + failure.percent);
            System.out.println("=====================================");
        }
        return answer;
    }

    private static class Failure implements Comparable<Failure> {
        float percent;
        int stage;

        public Failure(float percent, int stage) {
            this.percent = percent;
            this.stage = stage;
        }

        @Override
        public int compareTo(Failure f) {
            if (f.percent == this.percent) {
                return this.stage - f.stage;
            }
            return f.percent > this.percent ? 1 : -1;
        }
    }
}
